### PR TITLE
add state root tracking and refactor gasmeter -> state tracker

### PR DESCRIPTION
**Pull request by frrist**

### CODE REVIEW

File: box/box.go [hunk 1]
```diff
@@ -5,11 +5,11 @@
 package box
 
 type resourceBox struct {
```
Comment (frrist) [inline box/box.go:7]: resource box is now generic store, it contains `ApplyMessageResult`'s and `ApplyTipSetMessagesResult`s. When we want to add another value to track we can put it in one of the result structures.

File: drivers/test.go [hunk 2]
```diff
@@ -344,45 +344,53 @@ func (td *TestDriver) ApplyMessageSigned(msg *types.Message) (result chain.Apply
...
-		expectedGasUsed, ok := td.GasMeter.NextExpectedGas()
+		expectedGasUsed, ok := td.StateTracker.NextExpectedGas()
 		if ok {
 			assert.Equal(td.T, expectedGasUsed, result.Receipt.GasUsed, "Expected GasUsed: %d Actual GasUsed: %d", expectedGasUsed, result.Receipt.GasUsed)
 		} else {
 			foundGas = false
 		}
 	}
-	// TODO in the very near future we will be validating the stateroot here, keep in back of head.
+	if td.Config.ValidateStateRoot() {
```
Comment (frrist) [inline drivers/test.go:385]: The only meaningful change in this file to wire up stateroot validation.

File: tracker/track.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+	switch v := data.(type) {
+	case chain.ApplyMessageResult:
+		root, err := cid.Decode(v.Root)
+		if err != nil {
+			t.Fatal(err)
+		}
+		gasUsed = append(gasUsed, v.Receipt.GasUsed)
+		stateRoots = append(stateRoots, root)
+		return
+	case []chain.ApplyMessageResult:
```
Comment (anorth) [inline tracker/track.go:114]: The slice/single is annoying. Is there any way to get it to always be a slice?

File: chain/validator.go [hunk 4]
```diff
@@ -12,10 +12,16 @@ type Validator struct {
...
 
-type ApplyResult struct {
+type ApplyMessageResult struct {
 	Receipt types.MessageReceipt
 	Penalty abi.TokenAmount
 	Reward  abi.TokenAmount
+	Root    string
+}
+
+type ApplyTipSetMessagesResult struct {
```
Comment (anorth) [inline chain/validator.go]: ApplyTipSetResult is probably enough, and briefer.

**Review by anorth — APPROVED** [review]

File: tracker/track.go (hunk 3 above)
Comment (frrist) [reply tracker/track.go:114]: I suppose for the case that it _is_ a single could just shove it in a slice and reuse the same logic.

**Review by arajasek — APPROVED** [review]
Yup, looks good to me! Thanks!

**Pull request merged**